django/db/models/options.py

```diff
@@ -979,6 +979,18 @@ def _property_names(self):
                 names.append(name)
         return frozenset(names)
 
+    @cached_property
+    def _property_and_field_names(self):
+        """
+        Return a set of the names of the properties and fields defined on the model.
+        """
+        names = list(self._property_names)
+        for field in self._get_fields(reverse=False):
+            names.append(field.name)
+            if hasattr(field, "attname") and field.name != field.attname:
+                names.append(field.attname)
+        return frozenset(names)
```

Comment:
I'm not sure why we're adding properties here :thinking: As far as I'm aware, we should reuse `concrete_fields`, maybe in a new hook similar to the `_non_pk_concrete_field_names`, e.g.:
```python
    @cached_property
    def _concrete_field_names(self):
        names = []
        for field in self.concrete_fields
            names.append(field.name)
            if field.name != field.attname:
                names.append(field.attname)
        return frozenset(names)
```

Reply:
Since we can pass properties on `create()` if they have setters implemented, the idea was to include those as well to avoid raising the exception in those cases. Maybe there is a better way to achieve that 😅

Reply:
I also added a hook to have into consideration the generic foreign keys, but I think there is still one scenario that raises the exception when it shouldn't (related to ForeignObject fields)